### ISSUE TITLE
Refactor airshipctl config

### ISSUE THREAD

**Issue by kozhukalov** [airshipit/airshipctl#218]
**Problem description (if applicable)**
IMO, at the moment `airshipctl config` has some flaws, that could be fixed

1) Generating config files
1.1) If there is no config files in the ~/.airship directory, running any `airshipctl config` subcommand leads to generating the default config instead of suggesting to a user to generate default config using the `airshipctl config init` command and exit.  
1.2) Config reconcilation is too aggressive. If the config file is wrong, we generate the default config (related issue #176 ) 
1.3) We generate two files: 1. airshipctl config 2. kubectl config. IMO we should only take care of airshipctl config, but not of kubectl config. If k8s cluster already exists (operational), then the kubectl config also exists. If it is assumed to be deployed, then it is a matter of kubeadm/kind/etc. to generate the kubectl config. 
2) Subcommands
2.1) There are get-context and get-cluster subcommands which are similar to those of kubectl, but in kubectl those are plural get-contexts and get-clusters. If a user has experience with kubectl, it is better to align our commands with his experience.
2.2) In kubectl there are useful subcommands delete-cluster, delete-context, current-context, view, rename-context which are absent in airshipctl. It would be nice to have at least some of these subcommands in airshipctl.
3) Config structure
3.1) kind: Config should be changed into something like kind: AirConfig to avoid potential name conflict if at some point in the future configs will be aggregated somehow (like CR names)
3.2) kubectl configuration structure assumes all fields like contexts, users, clusters are arrays of maps, but in airshipctl all fields are just maps which might confuse experienced kubectl users.
 
**Proposed change**
1.1) If there is no airshipctl config file, suggest to use `airshipctl config init` and exit
1.2) If config is wrong, raise the error and exit
1.3) If there is no kubectl config, it is ok. The kubectl config will be generated by kubeadm/kind/etc. and we take it once it is ready. Another approach here could be just generate kubectl config file from manifests on our own since all necessary parameters (user/cluster names, certificates, etc) are supposed to be defined in manifests.
2.1) Align subcommand names with kubectl ones.
2.2) Implement delete-*, view, current-context commands (low priority)
3.1) Rename config resource into AirshipConfig
3.2) Use lists of maps in the config instead of maps. 

**Potential impacts**
No.

**Comment by dukov** [comment]
> 3.1) Rename config resource into AirshipConfig
GroupVersionKind should guarantee uniqueness of a resource. Same kind may lead to user confusion though. So I'd consider this as low priority as well

**Comment by kozhukalov** [comment]
> > GroupVersionKind should guarantee uniqueness of a resource. Same kind may lead to user confusion though. So I'd consider this as low priority as well
We don't serve configs via API server, so configs might be out of scope of this concern, but still when you extend the API, apiextensions-apiserver checks if there are kind conflicts. 

https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/controller/status/naming_controller.go#L174-L180

**Comment by kozhukalov** [comment]
Thanks @teoyaomiqui and @dukov. 
apiextensions-apiserver indeed considers resources names in the context of group https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/controller/status/naming_controller.go#L125 So then 3.1)  isn't actual and should be removed from the scope.

**Comment by kozhukalov** [comment]
Yet another issue with the config that confuses me personally (maybe other newbies as well might be confused) is that context in default Airship config ties together manifest and kubectl context 
```
contexts:
  default:
    contextKubeconf: default_target
    manifest: default
```
but bootstrapInfo, managementConfiguration and clusters are out of the scope of the context section (my kubectl experience tells me that context is something that ties together other sections in the conig).
Instead Airship config has clusters section that ties together bootstrapInfo, kubectl cluster (!? not context) and managementConfiguration
```
clusters:
  default:
    clusterType:
      target:
        bootstrapInfo: default
        clusterKubeconf: default_target
        managementConfiguration: default
```

**Comment by kozhukalov** [comment]
I would like to work on this. Please assign to me.

**Comment by kozhukalov** [comment]
Yet another issue:
If you have bootstapInfo section like this
```
bootstrapInfo:
  bootstrap-info-1:
    ...
  bootstrap-info-2:
    ...
```
the `rootSettings.Config` still contains `default` bootstrap subsection (which I might not need).

If the clusters section contains the link to nonexistent bootstrap info, like this for example
```
bootstrapInfo:
  bootstrap-info-1:
     ...
  bootstrap-info-2:
     ...
clusters:
  default:
    clusterType:
      target:
        bootstrapInfo: default
        ...
```
the `airshipctl config` should raise an error, not just silently add defaults

**Comment by ian-howell** [comment]
> 2.1) There are get-context and get-cluster subcommands which are similar to those of kubectl, but in kubectl those are plural get-contexts and get-clusters. If a user has experience with kubectl, it is better to align our commands with his experience.

This is implemented in https://review.opendev.org/#/c/725968/

**Comment by kozhukalov** [comment]
I am trying to summarize all the refactoring work here  https://etherpad.opendev.org/p/Airship_Config

**Comment by jezogwza** [comment]
This sparked part of the reasoning behind refactoring kubeconfig aspects into the document set. There are plenty if issues to capture the work for the refactoring will close this one,.
